hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieAvroHFileReader.java

```diff
@@ -684,6 +685,10 @@ private static class RecordIterator implements ClosableIterator<IndexedRecord> {
     public boolean hasNext() {
       try {
         // NOTE: This is required for idempotency
+        if (eof) {
+          return false;
+        }
```

Comment:
Seems a fundamental change, why the HFile reader can work now?

Reply:
Under what condition does the infinite iteration happen?  How to reproduce it in a test?

Reply:
This is being handled correctly in KeyPrefixIterator but is missing in this RecordIterator. The fix is to bring the change also to this class.